Assets/Mirror/Components/NetworkSceneChecker.cs

```diff
@@ -0,0 +1,119 @@
+using UnityEngine;
+using System.Collections.Generic;
+using UnityEngine.SceneManagement;
+
+namespace Mirror
+{
+    /// <summary>
+    /// Component that controls visibility of networked objects between scenes.
+    /// <para>Any object with this component on it will only be visible to other objects in the same scene</para>
+    /// <para>This would be used when the server has multiple additive subscenes loaded to isolate players to their respective subscenes</para>
+    /// </summary>
+    [AddComponentMenu("Network/NetworkSceneChecker")]
+    [RequireComponent(typeof(NetworkIdentity))]
+    [HelpURL("https://mirror-networking.com/docs/Components/NetworkSceneChecker.html")]
+    public class NetworkSceneChecker : NetworkBehaviour
+    {
+        /// <summary>
+        /// Flag to force this object to be hidden from all observers.
+        /// <para>If this object is a player object, it will not be hidden for that client.</para>
+        /// </summary>
+        [Tooltip("Enable to force this object to be hidden from all observers.")]
+        public bool forceHidden;
+
+        static Dictionary<string, HashSet<NetworkIdentity>> sceneCheckerObjects;
+
+        [Scene, SerializeField]
+        string currentScene;
+
+        [ServerCallback]
+        void Awake()
+        {
+            if (sceneCheckerObjects == null)
+                sceneCheckerObjects = new Dictionary<string, HashSet<NetworkIdentity>>();
+
+            currentScene = gameObject.scene.name;
+        }
+
+        public override void OnStartServer()
+        {
+            if (!sceneCheckerObjects.ContainsKey(currentScene))
+                sceneCheckerObjects.Add(currentScene, new HashSet<NetworkIdentity>());
+
+            sceneCheckerObjects[currentScene].Add(netIdentity);
+        }
+
+        [ServerCallback]
+        void Update()
+        {
+            if (currentScene == gameObject.scene.name)
+                return;
+
+            // This object is in a new scene so observers in the prior scene
+            // and the new scene need to rebuild their respective observers lists.
+
+            // Remove this object from the hashset of the scene it just left
+            sceneCheckerObjects[currentScene].Remove(netIdentity);
+
+            // RebuildObservers of all NetworkIdentity's in the scene this object just left
+            RebuildSceneObservers();
+
+            // Set this to the new scene this object just entered
+            currentScene = gameObject.scene.name;
+
+            // Make sure this new scene is in the dictionary
+            if (!sceneCheckerObjects.ContainsKey(currentScene))
+                sceneCheckerObjects.Add(currentScene, new HashSet<NetworkIdentity>());
+
+            // Add this object to the hashset of the new scene
+            sceneCheckerObjects[currentScene].Add(netIdentity);
+
+            // RebuildObservers of all NetworkIdentity's in the scene this object just entered
+            RebuildSceneObservers();
```

Comment:
why is this in here twice?

Reply:
before and after currentScene is changed